File: cache/q-1/lru.cpp
#include <unordered_map>
using namespace std;

class LRUCache {
private:
    struct Node {
        int key;
        int value;
        Node* prev;
        Node* next;
        Node(int k, int v) : key(k), value(v), prev(nullptr), next(nullptr) {}
    };

    int capacity;
    std::unordered_map<int, Node*> cache; // Maps key -> Node pointer
    Node* head; // head
    Node* tail; // tail

    // ----------------------------------------------
    // HELPER FUNCTIONS
    // ----------------------------------------------

    // #1. Helper function to add node right after head
    void addNode(Node* node) {
        node->next = head->next;
        node->prev = head;
        head->next->prev = node;
        head->next = node;
    }

    // #2. Helper function to remove existing node from list
    void removeNode(Node* node) {
        Node* prevNode = node->prev;
        Node* nextNode = node->next;
        prevNode->next = nextNode;
        nextNode->prev = prevNode;
    }

    /** #3. Helper function to move node to head,
    *   marks it as most recently used
    */
    void moveToHead(Node* node) {
        removeNode(node);
        addNode(node);
    }

    /** #4. Helper function to pop least recently used node,
    *   one before tail
    */
    Node* popTail() {
        Node* res = tail->prev;
        removeNode(res);
        return res;
    }

//////////////////////////////////////////////
/////////////////////////////////////////////
public:
    LRUCache(int capacity) : capacity(capacity) {
        // #1. Initialize head & tail
        head = new Node(0, 0);
        tail = new Node(0, 0);
        head->next = tail;
        tail->prev = head;
    }
    
    int get(int key) {
        if (cache.find(key) == cache.end()) {
            return -1;
        }
        // #2. If key exists, move it to head, return value
        Node* node = cache[key];
        moveToHead(node);
        return node->value;
    }
    
    void put(int key, int value) {
        if (cache.find(key) != cache.end()) {
            // #3. If key exists, update value, move to head
            Node* node = cache[key];
            node->value = value;
            moveToHead(node);
        } else {
            // #4. If key is new, create new node
            Node* newNode = new Node(key, value);
            cache[key] = newNode;
            addNode(newNode);

            // #5. If capacity exceeded, evict LRU item
            if (cache.size() > capacity) {
                Node* lru = popTail();
                cache.erase(lru->key);
                delete lru; // Free memory
            }
        }
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

File: cache/q-1/lru.ts
class Node {
    key: number;
    value: number;
    prev: Node | null;
    next: Node | null;

    constructor(key: number, value: number) {
        this.key = key;
        this.value = value;
        this.prev = null;
        this.next = null;
    }
}

class LRUCache {
    private capacity: number;
    private cache: Map<number, Node>;
    private head: Node; // head (MRU side)
    private tail: Node; // tail (LRU side)

    constructor(capacity: number) {
        this.capacity = capacity;
        this.cache = new Map<number, Node>();

        // Initialize nodes
        this.head = new Node(0, 0);
        this.tail = new Node(0, 0);
        this.head.next = this.tail;
        this.tail.prev = this.head;
    }

    get(key: number): number {
        const node = this.cache.get(key);
        if (!node) {
            return -1;
        }
        // Mark as recently used
        this.moveToHead(node);
        return node.value;
    }

    put(key: number, value: number): void {
        const existingNode = this.cache.get(key);

        if (existingNode) {
            // Update existing node
            existingNode.value = value;
            this.moveToHead(existingNode);
        } else {
            // Insert new node
            const newNode = new Node(key, value);
            this.cache.set(key, newNode);
            this.addNode(newNode);

            // Pop if capacity exceeded
            if (this.cache.size > this.capacity) {
                const lru = this.popTail();
                this.cache.delete(lru.key);
            }
        }
    }

    // --- Doubly Linked List Helper Methods ---

    private addNode(node: Node): void {
        // Insert right after head
        node.prev = this.head;
        node.next = this.head.next;

        if (this.head.next) {
            this.head.next.prev = node;
        }
        this.head.next = node;
    }

    private removeNode(node: Node): void {
        const prevNode = node.prev;
        const nextNode = node.next;

        if (prevNode) prevNode.next = nextNode;
        if (nextNode) nextNode.prev = prevNode;
    }

    private moveToHead(node: Node): void {
        this.removeNode(node);
        this.addNode(node);
    }

    private popTail(): Node {
        // LRU node, just before tail
        const res = this.tail.prev as Node;
        this.removeNode(res);
        return res;
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * var obj = new LRUCache(capacity)
 * var param_1 = obj.get(key)
 * obj.put(key,value)
 */